### PR TITLE
[sketch][check runs] Improve interpolation (compatible) METAGG-315

### PR DESCRIPTION
**Pull request by jbarciauskas**
### What does this PR do?

This is a backwards compatible version of #4887 that keeps uint16 buckets, resulting in much higher memory use but still greatly improves performance.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

```
➜ datadog-agent git:(joel.barciauskas/interp-uint16) /usr/local/bin/go test -benchmem -run=^$ github.com/DataDog/datadog-agent/pkg/aggregator -bench ^\(BenchmarkAddBucketWide1e10\)$
goos: darwin
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/aggregator
BenchmarkAddBucketWide1e10-4   	       4	 291443732 ns/op	1400899330 B/op	    1401 allocs/op
PASS
ok  	github.com/DataDog/datadog-agent/pkg/aggregator	3.082s
```

Note that this test does not run even once in 11 minutes on master.

Also, for comparison, the results from the uint32 branch (#4887 )

```
➜ datadog-agent git:(joel.barciauskas/better-interp-buckets) /usr/local/bin/go test -benchmem -run=^$ github.com/DataDog/datadog-agent/pkg/aggregator -bench ^\(BenchmarkAddBucketWide1e10\)$
goos: darwin
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/aggregator
BenchmarkAddBucketWide1e10-4   	    3612	    318575 ns/op	   76086 B/op	     412 allocs/op
PASS
ok  	github.com/DataDog/datadog-agent/pkg/aggregator	2.095s
```

### CODE REVIEW

**Review by olivielpeau — DISMISSED** [review]
Can't really review the logic in details as I'm not familiar with it (and no one in Core is), approving for Core in any case.

Once this is merged I guess we can merge https://github.com/DataDog/datadog-agent/pull/4867 as well, even though performance for values even slightly above 2^31 may still be too poor in practice.

File: pkg/aggregator/check_sampler.go [hunk 1]
```diff
@@ -107,35 +105,16 @@ func (cs *CheckSampler) addBucket(bucket *metrics.HistogramBucket) {
...
-		linearIncr = bucketRange / float64(bucket.Value)
-		countPerIncr = 1
-		incrCount = bucket.Value
-	}
 	if math.IsInf(bucket.UpperBound, 1) {
-		// We simulate the behavior of promQL for the infinity bucket:
-		// "if the quantile falls into the highest bucket, the upper bound of the 2nd highest bucket is returned"
-		incrCount = 1
-		countPerIncr = uint(bucket.Value)
+		cs.sketchMap.insertInterp(int64(bucket.Timestamp), contextKey, bucket.LowerBound, bucket.LowerBound, uint(bucket.Value))
```
Comment (mfpierre) [inline pkg/aggregator/check_sampler.go]: Could we add a comment on why we're doing this?
I had to double check what was the difference between the line bellow until I noticed the double `LowerBound` 😅

File: pkg/aggregator/sketch_map.go [hunk 2]
```diff
@@ -35,12 +35,16 @@ func (m sketchMap) insert(ts int64, ck ckey.ContextKey, v float64) bool {
 	return true
 }
 
-func (m sketchMap) insertN(ts int64, ck ckey.ContextKey, v float64, n uint) bool {
-	if math.IsInf(v, 0) || math.IsNaN(v) {
+func (m sketchMap) insertInterp(ts int64, ck ckey.ContextKey, l float64, u float64, n uint) bool {
```
Comment (mfpierre) [inline pkg/aggregator/sketch_map.go]: Could we add comments or more explicit variable names on this function? (especially for`l`ower & `u`pper)

File: pkg/metrics/test_helper.go [hunk 3]
```diff
@@ -70,8 +71,23 @@ func AssertSerieEqual(t *testing.T, expected, actual *Serie) {
...
 func AssertSketchSeriesEqual(t assert.TestingT, exp, act SketchSeries) {
+	assertSketchSeriesEqualWithComparator(t, exp, act, func(exp, act *quantile.Sketch) bool {
+		return exp.Equals(act)
+	})
+}
+
+// AssertSketchSeriesApproxEqual checks whether two SketchSeries are approximately equal
+func AssertSketchSeriesApproxEqual(t assert.TestingT, exp, act SketchSeries) {
+	assertSketchSeriesEqualWithComparator(t, exp, act, func(exp, act *quantile.Sketch) bool {
+		return quantile.SketchesApproxEqual(exp, act, .05)
```
Comment (mfpierre) [inline pkg/metrics/test_helper.go]: how to we determine the `.05` in this case? and if it's hardcoded should it be a parameter of `SketchesApproxEqual`

File: pkg/quantile/agent.go [hunk 4]
```diff
@@ -57,16 +61,32 @@ func (a *Agent) Insert(v float64) {
...
 }
 
-// InsertN inserts v, n times into the sketch.
-func (a *Agent) InsertN(v float64, n uint) {
-	a.Sketch.Basic.InsertN(v, n)
-
-	for i := 0; i < int(n); i++ {
-		a.Buf = append(a.Buf, agentConfig.key(v))
+// InsertInterpolate linearly interpolates counts from lower (l) to upper (u)
+func (a *Agent) InsertInterpolate(l float64, u float64, n uint) {
```
Comment (mfpierre) [inline pkg/quantile/agent.go]: nit: might be worth unit testing this even though it's covered by check_sampler tests

File: pkg/quantile/store_test.go [hunk 5]
```diff
@@ -159,8 +159,6 @@ func TestStore(t *testing.T) {
 			c("0:1 0:max 0:max", "0:3 0:max 0:max", 0, 0),
 			c("1:1 3:1 4:1 5:1 6:1 7:1", "1:1 2:1 3:2 4:1 5:1 6:1 7:1", 2, 3),
 			c("1:1 3:1", "1:1 2:3 3:1", 2, 2, 2),
-			c("1:1", "0:3 0:max 1:1", make([]Key, maxBinWidth+3)...),
```
Comment (mfpierre) [inline pkg/quantile/store_test.go]: why were these test cases removed?

**Review by mfpierre — COMMENTED** [review]
Looking good overall, left a few remarks and questions 👍

File: pkg/quantile/store_test.go (hunk 5 above)
Comment (jbarciauskas) [reply pkg/quantile/store_test.go]: These weren't feasible with uint32. even with uint16, I'm not sure what the point of making 2^16+3 keys is?

File: pkg/metrics/test_helper.go (hunk 3 above)
Comment (jbarciauskas) [reply pkg/metrics/test_helper.go]: This was experimental based on the actual data (I think the max error was like ~2%). I feel like it's useful to be able to choose this based on your test.

Comment (jbarciauskas) [reply pkg/metrics/test_helper.go]: Oh, I thought I had pushed it up. I can do that at least.

**Review by mfpierre — DISMISSED** [review]

**Review by mfpierre — DISMISSED** [review]
💯

**Review by mfpierre — DISMISSED** [review]

**Review by olivielpeau — DISMISSED** [review]

**Review by kenny-house — APPROVED** [review]
👍 Interpolation logic and `insertCounts` looks good to me

**Review by mfpierre — APPROVED** [review]

**Review by olivielpeau — APPROVED** [review]

**Pull request merged**